src/main/java/com/nimbusds/jose/crypto/impl/ECDHCryptoProvider.java

```diff
@@ -252,4 +247,79 @@ protected JWECryptoParts encryptWithZ(final JWEHeader header,
 	}
 
 
+	protected JWECryptoParts encryptMulti(final JWEHeader header,
+										  final List<Pair<UnprotectedHeader, SecretKey>> sharedSecrets,
+										  final byte[] clearText) throws JOSEException {
+
+		final ECDH.AlgorithmMode algMode = ECDH.resolveAlgorithmMode(header.getAlgorithm());
+		final SecretKey cek = ContentCryptoProvider.generateCEK(
+				header.getEncryptionMethod(),
+				getJCAContext().getSecureRandom()
+		);
+
+		List<Recipient> recipients = new ArrayList<>();
+		boolean encrypted = false;
+		JWECryptoParts parts = null;
+
+		for (Pair<UnprotectedHeader, SecretKey> rs : sharedSecrets) {
+			Base64URL encryptedKey = null;
+
+			if (!encrypted) {
+				parts = encryptWithZ(header, rs.getRight(), clearText, cek);
+				encryptedKey = parts.getEncryptedKey();
+				encrypted = true;
+			} else if (algMode.equals(ECDH.AlgorithmMode.KW)) {
+				SecretKey sharedKey = ECDH.deriveSharedKey(header, rs.getRight(), getConcatKDF());
+				encryptedKey = Base64URL.encode(AESKW.wrapCEK(cek, sharedKey, getJCAContext().getKeyEncryptionProvider()));
+			}
+
+			if (encryptedKey != null) {
+				recipients.add(new Recipient(rs.getLeft(), encryptedKey));
+			}
+		}
+
+		if (parts == null) {
+			throw new JOSEException("Content MUST be encrypted");
+		}
+
+		return new JWECryptoParts(
+				parts.getHeader(),
+				Collections.unmodifiableList(recipients),
+				parts.getInitializationVector(),
+				parts.getCipherText(),
+				parts.getAuthenticationTag()
+		);
+	}
+
+	protected byte[] decryptMulti(final JWEHeader header,
```

Comment:
This seems to be the exact same as the `decryptMulti` in `ECDH1PUCryptoProvider` - perhaps it can be extracted and reused?